src/editor/scroller.cpp

```diff
@@ -0,0 +1,159 @@
+//  SuperTux
+//  Copyright (C) 2016 Hume2 <teratux.mail@gmail.com>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include <math.h>
+
+#include "editor/scroller.hpp"
+
+#include "editor/editor.hpp"
+#include "supertux/colorscheme.hpp"
+#include "supertux/globals.hpp"
+#include "supertux/resources.hpp"
+#include "util/log.hpp"
+#include "video/drawing_context.hpp"
+#include "video/renderer.hpp"
+#include "video/video_system.hpp"
+
+namespace {
+
+const float TOPLEFT = 16;
+const float MIDDLE = 48;
+const float BOTTOMRIGHT = 80;
+const float SIZE = 96;
+
+}
+
+EditorScroller::EditorScroller() :
+  hidden(false),
+  scrolling(),
+  scrolling_vec(0, 0),
+  mouse_pos(0, 0)
+{
+}
+
+EditorScroller::~EditorScroller()
+{
+}
+
+bool
+EditorScroller::can_scroll() {
+  return scrolling && mouse_pos.x < SIZE && mouse_pos.y < SIZE;
+}
+
+void
+EditorScroller::draw(DrawingContext& context) {
+  if (hidden) return;
+
+  context.draw_filled_rect(Rectf(Vector(0, 0), Vector(SIZE, SIZE)),
+                           Color(0.9f, 0.9f, 1.0f, 0.6f),
+                           MIDDLE, LAYER_GUI-10);
+  if (can_scroll()) {
+    draw_arrow(context, mouse_pos);
+  }
+
+  draw_arrow(context, Vector(TOPLEFT, MIDDLE));
+  draw_arrow(context, Vector(BOTTOMRIGHT, MIDDLE));
+  draw_arrow(context, Vector(MIDDLE, TOPLEFT));
+  draw_arrow(context, Vector(MIDDLE, BOTTOMRIGHT));
+}
+
+void
+EditorScroller::draw_arrow(DrawingContext& context, Vector pos) {
+  Vector dir = pos - Vector(MIDDLE, MIDDLE);
+  if (dir.x != 0 || dir.y != 0) {
+    // draw a triangle
+    dir = dir.unit() * 8;
+    Vector dir2 = Vector(-dir.y, dir.x);
+    context.draw_triangle(pos + dir, pos - dir + dir2, pos - dir - dir2,
```

Comment:
This piece of code is responsible for the semi-transparent white surface which I am getting sometimes.